Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm

```diff
@@ -68,14 +73,82 @@
 void WebExtensionContext::permissionsRequest(HashSet<String> permissions, HashSet<String> origins, CompletionHandler<void(bool)>&& completionHandler)
 {
     HashSet<Ref<WebExtensionMatchPattern>> matchPatterns;
+
+    // If there is nothing to grant, return true. This matches Chrome and Firefox.
+    if (!permissions.size() && !origins.size()) {
+        firePermissionsEventListenerIfNecessary(_WKWebExtensionContextPermissionsWereGrantedNotification, permissions, matchPatterns);
+        completionHandler(true);
+        return;
+    }
+
     parseMatchPatterns(origins, matchPatterns);
 
     if (hasPermissions(permissions, matchPatterns)) {
         completionHandler(true);
         return;
     }
 
-    // FIXME: <https://webkit.org/b/250135> Call delegate method to prompt user for access.
+    // There shouldn't be any unsupported permissions, since they got reported as an error before this.
+    ASSERT(permissions.isSubset(extension().supportedPermissions()));
+
+    bool requestedAllHostsPattern = false;
+    for (auto& matchPattern : matchPatterns) {
+        if (matchPattern->matchesAllHosts()) {
+            requestedAllHostsPattern = true;
+            break;
+        }
+    }
+
+    if (requestedAllHostsPattern && !m_requestedOptionalAccessToAllHosts)
+        m_requestedOptionalAccessToAllHosts = YES;
+
+    __block MatchPatternSet grantedPatterns;
+    __block PermissionsSet grantedPermissions = permissions;
+    auto delegate = extensionController()->wrapper().delegate;
+
+    auto originsCompletionHandler = ^(NSSet<_WKWebExtensionMatchPattern *> *allowedOrigins) {
+        for (_WKWebExtensionMatchPattern *pattern in allowedOrigins) {
+            // Allowed origin doesn't match the origins requested.
+            if (!origins.contains(pattern.string)) {
+                completionHandler(false);
+                return;
+            }
+
+            grantedPatterns.add(*WebExtensionMatchPattern::getOrCreate(pattern.string));
+        }
+
+        grantPermissionMatchPatterns(WTFMove(grantedPatterns));
+        grantPermissions(WTFMove(grantedPermissions));
+        completionHandler(true);
```

Comment:
I put the call to `firePermissionsEventListenerIfNecessary()` in the `postAsyncNotification()` functions so that I would only have to call them in one place.